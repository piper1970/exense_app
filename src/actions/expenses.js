import uuid from 'uuid';
import  database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => (
    {
        type:'ADD_EXPENSE',
        expense
    }
);

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData;
        const expense = {
            description,note,amount,createdAt
        }
        return database.ref('expenses').push(expense)
        .then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        }).catch((e) => {
            console.log(`Error storing expense in database: ${e}`, expense);
        });
    };
};

// REMOVE _EXPENSE
export const removeExpense = ({id} = {}) => (
    {
        type:'REMOVE_EXPENSE',
        id
    }
);

export const startRemoveExpense = ({id} = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
        .then(() => {
            dispatch(removeExpense({id}))
        }).catch((error) => {
            console.log(`Errors occurred removing expense ${id} from database: ${error}`);
        });
    };
}; 

// SET_EXPENSES
export const setExpenses = (expenses) => {
    return {
        type:'SET_EXPENSES',
        expenses
    };
};

export const startSetExpenses = () => {
    const expenses = [];
    return (dispatch) => {
        return database.ref('expenses').once('value')
        .then((snapshot) =>{
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        }).catch((error) => {
            console.log(`Errors occurred setting expenses from database: ${error}`);
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => (
    {
        type:'EDIT_EXPENSE',
        id,
        updates
    }
);

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates)
        .then(() => {
            dispatch(editExpense(id, updates));
        }).catch((error) => {
            console.log(`Errors occurred updating expense ${id} on database: ${error}`);
        });
    }
}
