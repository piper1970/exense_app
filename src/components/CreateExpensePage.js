import React from 'react';

const CreateExpensePage = () => (
    <div>
        This is the create page!
    </div>
);

export default CreateExpensePage;