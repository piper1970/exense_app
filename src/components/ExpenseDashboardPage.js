import React from 'react';

const ExpenseDashboardPage = () => (
    <div>
        This is from the top of the world!
    </div>
);

export default ExpenseDashboardPage;