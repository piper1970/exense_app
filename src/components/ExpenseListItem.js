import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = (props) => {
    const {id, description, amount, note, createdAt, index} = props;

    return(
        <div>
            {isTextPrintable(description) && isValidId(id) && 
                <Link to={`/edit/${id}`}><h3>{description}</h3></Link>}
            {isAmountPrintable(amount) && <p>Amount: {numeral(amount/100).format('$0,0.00')}</p>}
            {isTextPrintable(note) && <p>Notes: {note}</p>}
            {isDatePrintable(createdAt) && 
                <p>Created at: {moment(createdAt).format('MMMM Do, YYYY')}</p>}
        </div>
            
        );
};


const isTextPrintable = (text) =>{
    return !!text && 
        typeof text === 'string' &&
        text.trim() !== 0;
};

const isDatePrintable = (date) => {
    return typeof date === 'number';
};

const isAmountPrintable = (amount) => {
    return !!amount &&
        typeof amount === 'number' &&
        amount !== 0;
}

const isValidId = (id) => {
    return !!id;
}

export default ExpenseListItem;