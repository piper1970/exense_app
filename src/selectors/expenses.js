const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = typeof text !== 'string' || 
            typeof expense.description !== 'string' ||
            expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {

        // desc order
        if(sortBy == 'date'){
            return a.createdAt < b.createdAt;
        }else{
            return a.amount < b.amount;
        }
    });
};

export default getVisibleExpenses;