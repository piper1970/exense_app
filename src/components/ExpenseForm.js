import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import  {removeExpense} from '../actions/expenses';
import {connect} from 'react-redux';

export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            id: props.expense?props.expense.id:undefined,
            description: props.expense? props.expense.description: '',
            note: props.expense? props.expense.note: '',
            amount: props.expense? (props.expense.amount/100).toString(): '',
            createdAt: props.expense? moment(props.expense.createdAt): moment(),
            calendarFocused: false,
            error:''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        if(this.state.description !== description){
            this.setState(() => ({description}));
        }
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        if(this.state.note !== note){
            this.setState(() => ({note}));
        }
    };

    amountChange = (e) => {
        const amount = e.target.value;
        const pattern = /^\d+(\.\d{0,2})?$/;
        if(!amount || amount.match(pattern)){
            this.setState(() => ({amount}));
        }
    };
    onDateChanged = (createdAt) => {
        if(!!createdAt){
            this.setState(() => ({createdAt}));
        }
    };
    onCalendarFocusChange = ({focused:calendarFocused}) => {
        this.setState(() => ({calendarFocused}));
    };

    onSubmitHandler = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            const error = 'Please provide description and amount';
            this.setState(() => ({error}));
        }else{
            const error = '';
            this.setState(() => ({error}));
            this.props.onSubmit({
                description:this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt:this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmitHandler}>
                    {!!this.state.error && <h3>{this.state.error}</h3>}
                    <div>
                        <label htmlFor="description">Description: </label>
                        <input 
                            type="text" 
                            name="description" 
                            placeholder="Description" 
                            value={this.state.description}
                            onChange={this.onDescriptionChange}
                            autoFocus />
                    </div>
                    <div>
                        <label htmlFor="amount">Amount: </label>
                        <input 
                            name="amount" 
                            type="text" 
                            placeholder="Amount"
                            value={this.state.amount}
                            onChange={this.amountChange}
                        />
                    </div>
                    <div>
                        <label htmlFor="datePicker">Choose a date: </label>
                        <SingleDatePicker
                            date={this.state.createdAt}
                            onDateChange={this.onDateChanged}
                            focused={this.state.calendarFocused}
                            onFocusChange={this.onCalendarFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                    <div>
                        <label htmlFor="notes">Notes: </label>
                        <textarea 
                            name="notes"
                            placeholder="Add an optional note for your expense..."
                            value={this.state.note}
                            onChange={this.onNoteChange}
                        />
                    </div>
                    <div>
                        <button>Expense</button>
                    </div>
                </form>
            </div>
        );  
    }
}

