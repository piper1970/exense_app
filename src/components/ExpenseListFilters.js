import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter} from '../actions/filters';

const ExpenseListFilters = (props) => (
    <div>
        <label htmlFor="filterTextBox">Add a text filter: </label>
        <input type="text" 
            value={props.filters.text} 
            name="filterTextBox" 
            placeholder="Add text here..."
            onChange={(e) => {
                props.dispatch(setTextFilter(e.target.value));
            }} />
    </div>
);

const connectMap = (state) => {
    if(true){
        return {
            filters:state.filters
        }
    }
};

export default connect(connectMap)(ExpenseListFilters);
