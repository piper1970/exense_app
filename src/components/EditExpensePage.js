import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
    return( 
        <div>
            <h1>Edit Expense</h1>
            <ExpenseForm
                expense = {props.expense}
                onSubmit={(expense) => {
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push('/');
                }}
            />
            <div>
                <button onClick={ () => {
                    props.dispatch(removeExpense({id:props.expense.id}));
                    props.history.push('/');
                    }
                }>Remove</button>
            </div>
        </div>
    );
};

const mapStateToForm = (state, props) => {
    const expense = state.expenses.find((expense) => expense.id === props.match.params.id);
    return {
        expense
    }
}
export default connect(mapStateToForm)(EditExpensePage);