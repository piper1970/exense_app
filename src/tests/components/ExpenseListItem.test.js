import {shallow} from 'enzyme';
import React from 'react';
import {ExpenseListItem} from '../../components/ExpenseListItem';
import moment from 'moment';
import expenses from '../fixtures/expenses';

const expense = expenses[0];

describe('ExpenseListItem', () => {

    test('should render ExpenseListItem when all props valid', () => {
        const wrapper = shallow(<ExpenseListItem 
            id={expense.id} 
            description={expense.description}
            amount={expense.amount}
            note={expense.note}
            createdAt={expense.createdAt}
            index={0}
        />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render expenselistitem without createdAt if not defined', () => {
        const wrapper = shallow(<ExpenseListItem 
            id={expense.id} 
            description={expense.description}
            amount={expense.amount}
            note={expense.note}
            index={0}
        />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render ExpenseListItem without note if not defined', () => {
        const wrapper = shallow(<ExpenseListItem 
            id={expense.id} 
            description={expense.description}
            amount={expense.amount}
            createdAt={expense.createdAt}
            index={0}
        />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render ExpenseListItem without amount if not defined', () => {
        const wrapper = shallow(<ExpenseListItem 
            id={expense.id} 
            description={expense.description}
            note={expense.note}
            createdAt={expense.createdAt}
            index={0}
        />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render ExpenseListItem without link if description not defined', () => {
        const wrapper = shallow(<ExpenseListItem 
            id={expense.id}
            amount={expense.amount}
            note={expense.note}
            createdAt={expense.createdAt}
            index={0}
        />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render ExpenseListItem without link if id not defined', () => {
        const wrapper = shallow(<ExpenseListItem
            description={expense.description}
            amount={expense.amount}
            note={expense.note}
            createdAt={expense.createdAt}
            index={0}
        />);
        expect(wrapper).toMatchSnapshot();
    });
});