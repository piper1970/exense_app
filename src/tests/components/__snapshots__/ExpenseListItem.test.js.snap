// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItem should render ExpenseListItem when all props valid 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    Amount: $
    1.95
  </p>
  <p>
    Notes: 
    note
  </p>
  <p>
    Created at: 
    Wednesday, December 31st, 1969
  </p>
</div>
`;

exports[`ExpenseListItem should render ExpenseListItem without amount if not defined 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    Notes: 
    note
  </p>
  <p>
    Created at: 
    Wednesday, December 31st, 1969
  </p>
</div>
`;

exports[`ExpenseListItem should render ExpenseListItem without link if description not defined 1`] = `
<div>
  <p>
    Amount: $
    1.95
  </p>
  <p>
    Notes: 
    note
  </p>
  <p>
    Created at: 
    Wednesday, December 31st, 1969
  </p>
</div>
`;

exports[`ExpenseListItem should render ExpenseListItem without link if id not defined 1`] = `
<div>
  <p>
    Amount: $
    1.95
  </p>
  <p>
    Notes: 
    note
  </p>
  <p>
    Created at: 
    Wednesday, December 31st, 1969
  </p>
</div>
`;

exports[`ExpenseListItem should render ExpenseListItem without note if not defined 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    Amount: $
    1.95
  </p>
  <p>
    Created at: 
    Wednesday, December 31st, 1969
  </p>
</div>
`;

exports[`ExpenseListItem should render expenselistitem without createdAt if not defined 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    Amount: $
    1.95
  </p>
  <p>
    Notes: 
    note
  </p>
</div>
`;
